### ISSUE TITLE
How to use a dataset not predefined in N2D2?

### ISSUE THREAD

**Issue by vankhoa21991** [CEA-LIST/N2D2#31]
Hello, 
For our project, I want to import my own dataset to N2D2.
It could be both images dataset or spike dataset from event camera like files .raw, .aedat, .dat, etc. 
Do you have any tutorial for that function. 
Thanks

**Comment by johannesthiele-cea** [comment]
Hello,

I updated the Environment class so that it is possible to either convert a static dataset or use a AER dataset. You can compare the models mnist-stdp.ini and mnist-stdp-NMNIST.ini to see how this is done for the MNIST and NMNIST dataset.s 

In general, you have to write your own driver if you want to include a dataset. A driver is a module that manages the training and test partition and how the dataset is read from the raw data. How this is done can be seen in the classes src/Database/MNIST_IDX_Database.cpp, src/Generator/MNIST_IDX_DatabaseGenerator.cpp (and the respective versions for NMNIST).

**Comment by vankhoa21991** [comment]
Thanks for your help, actually I can not find the models mnist-stdp.ini and mnist-stdp-NMNIST.ini.

**Comment by vankhoa21991** [comment]
Hello, I ran into this error while launching the mnist-stdpNMNIST.ini . I think the data is good, so it could be a problem in the code.

/n2d2 "mnist-stdp-NMNIST.ini" -learn 500000  -log 10000...
Option -log: number of steps between logs [10000]
Option -learn: number of backprop learning steps [500000]
Loading network configuration file mnist-stdp-NMNIST.ini
Layer: fc1 [Fc(Spike_Analog)]
Notice: Could not open configuration file: fc1.cfg
  # Synapses: 115650
  # Inputs dims: 34 34 2 
  # Outputs dims: 1 1 50 
Target: fc1 (target value: 1 / default value: 0 / top-n value: 1)
Total number of neurons: 50
Total number of nodes: 50
Total number of synapses: 115650
Total number of virtual synapses: 115650
Total number of connections: 115650
Notice: Unused section fc1.Target in INI file
Notice: Unused section common.config in INI file
Notice: Unused section common_Spike_RRAM.config in INI file
Learning database size: 30000 images
Validation database size: 30000 images
Testing database size: 10000 images
[LOG] Stimuli transformations flow (transformations.png)
sh: 1: dot: not found
[LOG] Network graph (mnist-stdp-NMNIST.ini.png)
[LOG] Network SVG graph (mnist-stdp-NMNIST.ini.svg)
sh: 1: dot: not found
[LOG] Network stats (stats/*)
[LOG] Solvers scheduling (schedule/*)
[LOG] Layer's receptive fields (receptive_fields.log)
[LOG] Labels mapping (*.Target/labels_mapping.log)
[LOG] Labels legend (*.Target/labels_legend.png)
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
         line 0: Can't plot with an empty x range!

Invalid registrar key "bin"
terminate called after throwing an instance of 'std::bad_function_call'
  what():  bad_function_call
10:12:25: The program has unexpectedly finished.

**Comment by vankhoa21991** [comment]
I found the answer, have to use -learn-stdp instead.

**Comment by vankhoa21991** [comment]
Hello, I tried to run the example but the performance is quite low ~11% success rate. And I notice an error on getTargetCell.


../../Debug2/bin/n2d2 ../../../models/mnist-stdp-NMNIST.ini -learn-stdp 100000 -log 10000
Option -log: number of steps between logs [10000]
Option -learn-stdp: number of STDP learning steps [100000]
Loading network configuration file ../../../models/mnist-stdp-NMNIST.ini
Layer: fc1 [Fc(Spike_Analog)]
Notice: Could not open configuration file: fc1.cfg
  # Synapses: 115650
  # Inputs dims: 34 34 2 
  # Outputs dims: 1 1 50 
Target: fc1 (target value: 1 / default value: 0 / top-n value: 1)
Total number of neurons: 50
Total number of nodes: 50
Total number of synapses: 115650
Total number of virtual synapses: 115650
Total number of connections: 115650
Notice: Unused section fc1.Target in INI file
Notice: Unused section common.config in INI file
Notice: Unused section common_Spike_RRAM.config in INI file
Learning database size: 30000 images
Validation database size: 30000 images
Testing database size: 10000 images
[LOG] Stimuli transformations flow (transformations.png)
sh: 1: dot: not found
[LOG] Network graph (mnist-stdp-NMNIST.ini.png)
sh: 1: [LOG] Network SVG graph (mnist-stdp-NMNIST.ini.svg)
dot: not found
[LOG] Network stats (stats/*)
[LOG] Solvers scheduling (schedule/*)
[LOG] Layer's receptive fields (receptive_fields.log)
[LOG] Labels mapping (*.Target/labels_mapping.log)
[LOG] Labels legend (*.Target/labels_legend.png)
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
         line 0: Can't plot with an empty x range!

Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty y range [115650:115650], adjusting to [114494:116806]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty cb range [115650:115650], adjusting to [114494:116806]
**Error for testing: DeepNet::getTargetCell(): wrong cell type for index 0**
Continue...
Learn from 1e-06s to 1e-06s...
   env: 4732   fc1: 61   (total: 4793)
   #4082 / success rate: 0%   [FAIL]
Learn from 1e-06s to 2e-06s...
   env: 3474   fc1: 43   (total: 3517)

**Comment by vankhoa21991** [comment]
Hello, I tried this example, and the acc reach ~72%
/n2d2 ../../../models/mnist-stdp.ini -learn-stdp 100000 -log 10000

But when I want to test the model, it demain the folder 'weights' and 'weights_range_normalized', when I only have the weights-init and weights-stdp instead.
Do you have any idea about this? Thanks

**Comment by davidbriand-cea** [comment]
Hello,
I think you have to specify the folder "weights-stdp" for testing : 
./n2d2 ../../../models/mnist-stdp.ini -w weights-stdp -test

**Comment by vankhoa21991** [comment]
Hello, it demands always the weights_range directory

/n2d2 ../../../models/mnist-stdp.ini -w weights-stdp -test
Option -test: perform testing
Option -w: start with weights imported from a specified location (even when loading a previously saved state) [weights-stdp]
Loading network configuration file ../../../models/mnist-stdp.ini
Layer: fc1 [Fc(Spike_Analog)]
Notice: Could not open configuration file: fc1.cfg
  # Synapses: 39250
  # Inputs dims: 28 28 1 
  # Outputs dims: 1 1 50 
Target: fc1 (target value: 1 / default value: 0 / top-n value: 1)
Total number of neurons: 50
Total number of nodes: 50
Total number of synapses: 39250
Total number of virtual synapses: 39250
Total number of connections: 39250
Notice: Unused section fc1.Target in INI file
Notice: Unused section common.config in INI file
Notice: Unused section common_Spike_RRAM.config in INI file
Learning database size: 60000 images
Validation database size: 0 images
Testing database size: 10000 images
[LOG] Stimuli transformations flow (transformations.png)
[LOG] Network graph (mnist-stdp.ini.png)
sh: 1: dot: not found
[LOG] Network SVG graph (mnist-stdp.ini.svg)
sh: 1: dot: not found
[LOG] Network stats (stats/*)
[LOG] Solvers scheduling (schedule/*)
[LOG] Layer's receptive fields (receptive_fields.log)
[LOG] Labels mapping (*.Target/labels_mapping.log)
[LOG] Labels legend (*.Target/labels_legend.png)
Warning: empty y range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty y range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty cb range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty cb range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty cb range [39250:39250], adjusting to [38857.5:39642.5]
Warning: empty y range [100:100], adjusting to [99:101]
Warning: empty cb range [100:100], adjusting to [99:101]
Warning: empty y range [50:50], adjusting to [49.5:50.5]
Warning: empty cb range [50:50], adjusting to [49.5:50.5]
         line 0: Can't plot with an empty x range!

Importing weights from directory 'weights-stdp'.
Notice: Could not open synaptic file: weights-stdp/fc1_weights.syntxt
[LOG] Learn frame samples (frames/frame*)
Notice: stimuli depth is 8U (according to database first stimulus)
[LOG] Test frame samples (frames/test_frame*)
Error for testing: DeepNet::getTargetCell(): wrong cell type for index 0
Continue...
Importing weights from directory 'weights_range_normalized'.
Time elapsed: 1.40786 s
Error: Could not open synaptic file: weights_range_normalized/fc1_weights.syntxt

**Comment by vankhoa21991** [comment]
Hello, I'm trying an experiment like training an ANN (1 layer FC) with MNIST, and then import the weights to a SNN with same structure to do inference on NMNIST. But there are two many options in the ini file that I don't understand, like should I choose the base model as Transcode, or as Spike_Analog. I post the two ini file here, could you help me to check it.

the training ini:

; Command:
; ./n2d2.sh "$N2D2_MODELS/LeNet.ini" -learn 6000000 -log 100000

DefaultModel=Frame_CUDA

; Database
[database]
Type=MNIST_IDX_Database
Validation=0.2 ; Use 20% of the dataset for validation

; Environment
[sp]
SizeX=32
SizeY=32
BatchSize=128

[sp.Transformation_1]
Type=ChannelExtractionTransformation
CSChannel=Gray

[sp.Transformation_2]
Type=RescaleTransformation
Width=[sp]SizeX
Height=[sp]SizeY

; Output layer (fully connected)
[fc1]
Input=sp
Type=Fc
NbOutputs=500
ActivationFunction=Rectifier
WeightsFiller=HeFiller
ConfigSection=common.config

[fc2]
Input=fc1
Type=Fc
NbOutputs=10
ActivationFunction=Linear
ConfigSection=common.config

[softmax]
Input=fc2
Type=Softmax
NbOutputs=[fc2]NbOutputs
WithLoss=1

[softmax.Target]

; Common config for static model
[common.config]
NoBias=0
WeightsSolver.LearningRate=0.05
WeightsSolver.Momentum=0.9
WeightsSolver.Decay=0.0005
Solvers.LearningRatePolicy=StepDecay
Solvers.LearningRateStepSize=[sp]_EpochSize
Solvers.LearningRateDecay=0.993
Solvers.Clamping=0.0:1.0

The test ini:

DefaultModel=Transcode_CUDA

; Database
[database]
Type=N_MNIST_Database

; Environment
[env]
SizeX=32
SizeY=32
BatchSize=1
NbChannels=2
ConfigSection=env.config

[env.config]
ReadAerData=1

[env.Transformation_1]
Type=RescaleTransformation
Width=[env]SizeX
Height=[env]SizeY

; Output layer (fully connected)
[fc1]
Input=env
Type=Fc
NbOutputs=500
ConfigSection=common.config
ConfigSection(Spike_Analog)=common_Spike_Analog.config
ConfigSection(Spike_RRAM)=common_Spike_RRAM.config

[fc2]
Input=fc1
Type=Fc
NbOutputs=10
ConfigSection=common.config,fc2.config
ConfigSection(Spike_Analog)=common_Spike_Analog.config
ConfigSection(Spike_RRAM)=common_Spike_RRAM.config

[fc2.Target]

[fc2.config]
; Spike-based computing
TerminateDelta=4
BipolarThreshold=1

; Common config for static model
[common.config]
NoBias=0
WeightsSolver.LearningRate=0.005
Threshold=1.0
Refractory=10,000,000

; Common config for Spike_Analog model
[common_Spike_Analog.config]
EnableStdp=0
Threshold=50.0
StdpLtp=50,000,000
InhibitRefractory=1,000
Leak=100,000,000
BipolarIntegration=0
RefractoryIntegration=0
WeightsMinMean="1.0e-4; 10%"
WeightsMaxMean="1.0; 10%"
WeightsRelInit="0.67; 20%"
WeightIncrement="0.0017; 1%"
WeightIncrementDamping=-3
WeightDecrement="0.0007; 1%"
WeightDecrementDamping=0

; Common config for Spike_RRAM model
[common_Spike_RRAM.config]
EnableStdp=0
Threshold=1.52e-2
LtdProba=0.0007
LtpProba=0.0020
StdpLtp=50,000,000
InhibitRefractory=1,000
Leak=100,000,000
BipolarWeights=0
BipolarIntegration=0
RefractoryIntegration=0
WeightsMinMean="1.19e-5; 1.67e-5"
WeightsMaxMean="3.04e-4; 3.86e-5"
WeightsRelInit="2.0e-4; 20%"
WeightsMaxVarSlope=-1.7575
WeightsMaxVarOrigin=-17.1691
WeightsMinVarSlope=-0.0791
WeightsMinVarOrigin=-1.7081
WeightsSetProba="0.843; 0.197"
WeightsResetProba="0.992; 0.00748"
SynapticRedundancy=64

Thanks

**Comment by vankhoa21991** [comment]
Actually I get the points, the weights_range_normalized folder only created by the function learn-stdp and test (image). So for testing an SNN trained by back propagation on ANN and run on pure spike (NMNIST), I guess it's not possible. Another question about the mnist-stdp-NMNIST, I can only reach ~12%, do you know about this?

**Comment by olivierbichler-cea** [comment]
Hi, there is a conflict between the -w parameter and the weight_range_normalized folder which is forced in testing when not using -learn-stdp. This behavior is confusing and must be changed. As a temporary fix, you can juste rename the weight_stdp or weight_validation folder into weight_range_normalized.
